airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java

```diff
@@ -284,17 +293,34 @@ public List<AutoCloseableIterator<AirbyteMessage>> getIncrementalIterators(final
       final Duration firstRecordWaitTime = PostgresUtils.getFirstRecordWaitTime(sourceConfig);
       LOGGER.info("First record waiting time: {} seconds", firstRecordWaitTime.getSeconds());
 
+      final PostgresDebeziumStateUtil postgresDebeziumStateUtil = new PostgresDebeziumStateUtil();
+      final JsonNode state =
+          (stateManager.getCdcStateManager().getCdcState() == null || stateManager.getCdcStateManager().getCdcState().getState() == null) ? null
+              : Jsons.clone(stateManager.getCdcStateManager().getCdcState().getState());
+      final boolean savedOffsetAfterReplicationSlotLSN = postgresDebeziumStateUtil.isSavedOffsetAfterReplicationSlotLSN(
+          Jsons.clone(PostgresCdcProperties.getDebeziumDefaultProperties(sourceConfig)),
+          catalog,
+          state,
+          // We can assume that there will be only 1 replication slot cause before the sync starts for Postgres CDC,
+          // we run all the check operations and one of the check validates that the replication slot exists and has only 1 entry
+          getReplicationSlot(database, sourceConfig).get(0),
+          sourceConfig);
+
+      if (!savedOffsetAfterReplicationSlotLSN) {
+        LOGGER.warn("Saved offset is before Replication slot's confirmed_flush_lsn, Airbyte will trigger sync from scratch");
+      }
+
       final AirbyteDebeziumHandler handler = new AirbyteDebeziumHandler(sourceConfig,
           PostgresCdcTargetPosition.targetPosition(database), false, firstRecordWaitTime);
       final PostgresCdcStateHandler postgresCdcStateHandler = new PostgresCdcStateHandler(stateManager);
       final List<ConfiguredAirbyteStream> streamsToSnapshot = identifyStreamsToSnapshot(catalog, stateManager);
       final Supplier<AutoCloseableIterator<AirbyteMessage>> incrementalIteratorSupplier = () -> handler.getIncrementalIterators(catalog,
-          new PostgresCdcSavedInfoFetcher(stateManager.getCdcStateManager().getCdcState()),
+          new PostgresCdcSavedInfoFetcher(savedOffsetAfterReplicationSlotLSN ? stateManager.getCdcStateManager().getCdcState() : null),
```

Comment:
When `savedOffsetAfterReplicationSlotLSN` is `false`, the data should be synced from scratch. However, it is not obvious to me which line of code is triggering this behavior. We are not creating new snapshot iterators when `savedOffsetAfterReplicationSlotLSN` is `false`.

Is that the incremental iterator constructed here will sync everything when the offset does not exist?

Reply:
Yup. The incremental iterator will sync everything from scratch cause debezium will look at the state and see it's null